packages/mask/src/plugins/Referral/SNSAdaptor/BuyToFarm.tsx

```diff
@@ -0,0 +1,228 @@
+    const { setDialog: openSwapDialog } = useRemoteControlledDialog(PluginTraderMessages.swapDialogUpdated)
+
+    // fetch all farms
+    const { value: farms = [] } = useAsync(
+        async () => farmsService.getAllFarms(currentChainId, ERC20),
+        [ERC20, currentChainId],
+    )
+
+    const onClickTokenSelect = useCallback(() => {
+        setSelectTokenDialog({
+            open: true,
+            uuid: id,
+            title: t('plugin_referral_select_a_token_to_buy_and_hold'),
+            onlyFarmTokens: true,
+        })
+    }, [id, setToken])
+
+    const swapToken = useCallback(() => {
+        if (!token) {
+            showSnackbar(t('plugin_referral_error_token_not_select'), { variant: 'error' })
+            return
+        }
+        openSwapDialog({
+            open: true,
+            traderProps: {
+                chainId: SWAP_CHAIN_ID,
+                coin: {
+                    id: token?.address,
+                    name: token?.name ?? '',
+                    symbol: token?.symbol ?? '',
+                    contract_address: token?.address,
+                    decimals: token?.decimals,
+                } as Coin,
+            },
+        })
+    }, [token, openSwapDialog])
+
+    const onConfirmReferFarm = useCallback(() => {
+        props?.onChangePage?.(PagesType.TRANSACTION, t('plugin_referral_transaction'), {
+            hideAttrLogo: true,
+            hideBackBtn: true,
+            transactionDialog: {
+                transaction: {
+                    status: TransactionStatus.CONFIRMATION,
+                    title: t('plugin_referral_transaction_complete_signature_request'),
+                    subtitle: t('plugin_referral_transaction_sign_the_message_to_register_address_for_rewards'),
+                },
+            },
+        })
+    }, [props])
+
+    const onError = useCallback(
+        (error?: string) => {
+            showSnackbar(error || t('go_wrong'), { variant: 'error' })
+            props?.onChangePage?.(PagesType.BUY_TO_FARM, `${TabsReferralFarms.TOKENS}: ${PagesType.BUY_TO_FARM}`)
+        },
+        [props],
+    )
+
+    const onClickBuyToFarm = useCallback(async () => {
+        if (!token?.address) {
+            return onError(t('plugin_referral_error_token_not_select'))
+        }
+        try {
+            onConfirmReferFarm()
+            await proofOfRecommendationService.singAndPostProofOfRecommendationWithReferrer(
+                web3,
+                account,
+                token.address,
+                MASK_REFERRER,
+            )
+            props?.onChangePage?.(PagesType.BUY_TO_FARM, `${TabsReferralFarms.TOKENS}: ${PagesType.BUY_TO_FARM}`)
+            swapToken()
+        } catch (error: any) {
+            onError(error?.message)
+        }
+    }, [web3, account, token, props])
+
+    const referredTokenFarms = token
+        ? farms.filter((farm) => farm.referredTokenDefn === toChainAddressEthers(token.chainId, token.address))
+        : []
```

Comment:
```suggestion
        : EMPTY_LIST
```

Reply:
TODO: add `no-empty-collection` rule to `@dimeniondev/eslint-plugin`